File: rust/introductory/creating_strings.rs
use std::io;

fn solve(cnt: &mut [u32; 30], size: usize, curr: String) {
    if size == 0 {
        println!("{}", curr);
        return;
    }
    for i in 0..30 {
        if cnt[i as usize] > 0 {
            cnt[i as usize] -= 1;
            solve(
                cnt,
                size - 1,
                format!("{}{}", curr, (i + 'a' as u8) as char),
            );
            cnt[i as usize] += 1;
        }
    }
}

fn fact(n: u32) -> u32 {
    if n == 0 {
        1
    } else {
        n * fact(n - 1)
    }
}

fn main() {
    let mut buffer = String::new();
    let mut cnt = [0; 30];

    io::stdin().read_line(&mut buffer);
    let buffer = buffer.trim();

    for c in buffer.as_bytes() {
        cnt[(c - 'a' as u8) as usize] += 1;
    }
    let mut ans: u32 = fact(buffer.len() as u32);
    for i in 0..30 {
        ans /= fact(cnt[i as usize]);
    }
    println!("{ans}");
    solve(&mut cnt, buffer.len(), String::new());
}

File: cpp/introductory/creating_strings.cpp
#include <bits/stdc++.h>

using namespace std;

#define endl '\n'
#define f1 first
#define s2 second
#define pb push_back
#define mp make_pair
#define int long long
#define fri(a) freopen(a,"r",stdin);
#define fro(a) freopen(a,"w",stdout);
const int MOD = 1e9 + 7;
const int N = 2e5 + 5;

int cnt[30];
string s;

int fact(int num) {
    if (num == 0) {
        return 1;
    }
    return num * fact(num - 1);
}

void solve(string curr) {
    if (curr.size() == s.size()) {
        cout << curr << endl;
        return;
    }

    for (int i = 0; i < 30; i++) {
        if (cnt[i] > 0) {
            cnt[i] -= 1;
            solve(curr + (char)(i + 'a'));
            cnt[i] += 1;
        }
    }
}

int32_t main() {

    //fri("in.txt");
    //fro("out.txt");

    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    cin >> s;

    for (auto& c : s) {
        cnt[c - 'a']++;
    }
    int ans = fact(s.size());
    for (int i = 0; i < 30; i++) {
        ans /= fact(cnt[i]);
    }
    cout << ans << endl;
    solve("");


    return 0;
}
